src/lsp.ts

```diff
@@ -0,0 +1,180 @@
+            vscode.window.showErrorMessage(`Bad hash: Expected ${expected}, got ${str}!`);
+            fs.unlinkSync(tmpPath);
+            return;
+          }
+          const zip = new Admzip(tmpPath);
+          zip.extractAllTo(lspDir);
+          if (os.platform() != "win32")
+            fs.chmodSync(path.join(lspDir, this.executableName), 0o755);
+          vscode.window.showInformationMessage("Language server is setup correctly!");
+          fs.unlinkSync(tmpPath);
+          this.restart();
+        }).catch((err: Error) => {
+          vscode.window.showErrorMessage(err.message);
+          fs.unlinkSync(tmpPath);
+        });
+      }).catch((err: Error) => {
+        vscode.window.showErrorMessage(err.message);
+        fs.unlinkSync(tmpPath);
+      });
+
+    } catch (err) {
+      vscode.window.showErrorMessage(JSON.stringify(err));
+      fs.unlinkSync(tmpPath);
+      return;
+    }
+  }
+
+  async computeFileHash(filePath: string): Promise<string> {
+    const hash = crypto.createHash("sha256");
+    const stream = fs.createReadStream(filePath);
+    stream.on("data", (data) => {
+      hash.update(data);
+    });
+    return new Promise<string>((resolve, reject) => {
+      stream.on("error", reject);
+      stream.on("end", () => {
+        resolve(hash.digest("hex"));
+      });
+    });
+  }
+
+  async downloadFile(url: string, dest: string): Promise<void> {
+    return new Promise((resolve, reject) => {
+      const file = fs.createWriteStream(dest);
+      const request = https.get(url, (response) => {
+        if (response.statusCode === 302 || response.statusCode === 301) {
+          const redirectUrl = response.headers.location;
+          this.downloadFile(redirectUrl, dest)
+            .then(() => resolve())
+            .catch((err) => reject(err));
+        } else {
+          response.pipe(file);
+          file.on('finish', () => {
+            file.close();
+            resolve();
+          });
+          file.on('error', (err) => {
+            vscode.window.showErrorMessage(`Error writing to file: ${err}`);
+            reject(err);
+          });
+        }
+      });
+      request.on('error', (err) => {
+        vscode.window.showErrorMessage(`Request error: ${err}`);
+        reject(err);
+      });
+    });
+  }
+
+  private findLanguageServer(): string | null {
+    const platform = os.platform();
+    const arch = os.arch();
+    if (!this.supportsSystem(platform, arch))
+      return null
+    const executableName = this.executableName;
+    const lspDir = path.join(this.getExtensionLSPDir(this.context), this.executableName);
+    const suffix = platform == "win32" ? ".exe" : "";
+    const fullpath = path.join(lspDir, executableName + suffix);
+    if (fs.existsSync(fullpath)) {
+      return fullpath;
+    }
+    return null;
+  }
+
+  private getExtensionLSPDir(ctx: vscode.ExtensionContext): string {
+    return path.join(ctx.globalStorageUri.fsPath, "lsp");
+  }
+
+  public dispose() {
```

Comment:
this should be made private

Reply:
I got this error:
```
src/extension.ts:292:26 - error TS2345: Argument of type 'SwiftMesonLspLanguageClient' is not assignable to parameter of type '{ dispose(): any; }'.
  Property 'dispose' is private in type 'SwiftMesonLspLanguageClient' but not in type '{ dispose(): any; }'.

292   ctx.subscriptions.push(client)
                             ~~~~~~


Found 1 error in src/extension.ts:292
```
After changing to private. Just having `dispose () {` seems to compile.

Reply:
Oh, I guess it is required by VSCode. Nevermind me :)